zebra-state/src/service/check/anchors.rs

```diff
@@ -47,26 +48,28 @@ pub(crate) fn sapling_orchard_anchors_refer_to_final_treestates(
                 ?anchor,
                 ?anchor_index_in_tx,
                 ?tx_index_in_block,
-                height = ?prepared.height,
+                ?height,
                 "observed sapling anchor",
             );
 
-            if !parent_chain.sapling_anchors.contains(&anchor)
+            if !parent_chain
+                .map(|chain| chain.sapling_anchors.contains(&anchor))
+                .unwrap_or(false)
                 && !finalized_state.contains_sapling_anchor(&anchor)
             {
                 return Err(ValidateContextError::UnknownSaplingAnchor {
                     anchor,
-                    height: prepared.height,
+                    height,
                     tx_index_in_block,
-                    transaction_hash: prepared.transaction_hashes[tx_index_in_block],
+                    transaction_hash: transaction.hash(),
```

Comment:
Repeatedly hashing the serialized bytes of large transactions is expensive, so calling `transaction.hash()` here will hurt performance.

Instead, we could pass the `UnminedTx` to the state request, and use `transaction.id.mined_id()` here.

This change is needed to all the `hash()` calls in this PR.

Reply:
I was a little concerned about that too but hadn't noticed that the mempool requests had the transaction hash already and was thinking errors happens occasionally enough that it wouldn't be too important for blocks.

Updated in [`55232d1`](https://github.com/ZcashFoundation/zebra/pull/5716/commits/55232d14135a0e82ae17d1c150223dbf4ded989d).

Reply:
Yeah unfortunately this is a denial of service risk, if a remote node can send failing mempool transactions, and get us to hash a lot of transaction data.

Mempool transactions aren't protected by proof of work, so we have to be careful about performance.